From e285d844e65d16698ece2917b128bda084abb919 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Maxime=20M=C3=A9r=C3=A9?= <maxime.mere@foss.st.com>
Date: Tue, 23 Jan 2024 11:24:58 +0100
Subject: [PATCH 465/954] feat(stm32mp1): derivating MCE key from HUK
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

This patch changes how the MCE master key is generated. Instead of using
a random key, we use a derivative key from the SAES hardware key. We
just have to store the 32-bit random seed in the backup context instead
of the full key.

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/arm-trusted-firmware.git 827f302bcb4670e8a14438e3edd410057a2d59d2 feat(stm32mp1): derivating MCE key from HUK]
Signed-off-by: Maxime Méré <maxime.mere@foss.st.com>
Change-Id: I5a3253d7d78929540b9d2a4d17fc991f6fe4340b
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/tf-a/+/353587
Reviewed-by: Yann GAUTIER <yann.gautier@foss.st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Domain-Review: Yann GAUTIER <yann.gautier@foss.st.com>
---
 plat/st/stm32mp1/bl2_plat_setup.c           | 43 ++++++++++++++++++---
 plat/st/stm32mp1/include/stm32mp1_context.h |  4 +-
 plat/st/stm32mp1/platform.mk                |  4 +-
 plat/st/stm32mp1/stm32mp1_context.c         | 10 ++---
 plat/st/stm32mp1/stm32mp1_def.h             |  1 +
 5 files changed, 47 insertions(+), 15 deletions(-)

diff --git a/plat/st/stm32mp1/bl2_plat_setup.c b/plat/st/stm32mp1/bl2_plat_setup.c
index 5ef22b3e4e..e82f932a55 100644
--- a/plat/st/stm32mp1/bl2_plat_setup.c
+++ b/plat/st/stm32mp1/bl2_plat_setup.c
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2015-2023, Arm Limited and Contributors. All rights reserved.
+ * Copyright (c) 2015-2024, Arm Limited and Contributors. All rights reserved.
  *
  * SPDX-License-Identifier: BSD-3-Clause
  */
@@ -24,6 +24,9 @@
 #include <drivers/st/stm32_mce.h>
 #endif
 #include <drivers/st/stm32_rng.h>
+#if STM32MP13
+#include <drivers/st/stm32_saes.h>
+#endif
 #include <drivers/st/stm32_uart.h>
 #include <drivers/st/stm32mp1_clk.h>
 #include <drivers/st/stm32mp1_pwr.h>
@@ -457,30 +460,58 @@ skip_console_init:
 }
 
 #if STM32MP13
+static int generate_mce_key_from_seed(uint8_t *mkey, uint8_t *seed, uint32_t seed_size){
+	int ret;
+	struct stm32_saes_context ctx;
+	uint8_t payload[MCE_KEY_SIZE_IN_BYTES] = {0U};
+
+	assert(seed_size <= MCE_KEY_SIZE_IN_BYTES);
+
+	ret = stm32_saes_driver_init();
+	if (ret != 0) {
+		return ret;
+	}
+
+	memcpy(payload, seed, seed_size); /* add seed in a block sized payload */
+
+	ret = stm32_saes_init(&ctx, false, STM32_SAES_MODE_ECB,
+			      STM32_SAES_KEY_DHU, NULL, MCE_KEY_SIZE_IN_BYTES, NULL, 0U);
+	if (ret != 0) {
+		return ret;
+	}
+
+	return stm32_saes_update(&ctx, true, payload, mkey, sizeof(payload));
+}
+
 static void prepare_encryption(void)
 {
 	uint8_t mkey[MCE_KEY_SIZE_IN_BYTES];
+	uint8_t seed[MCE_SEED_SIZE_IN_BYTES];
 
 	stm32_mce_init();
 
 #if STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER
-	if (stm32_rng_read(mkey, MCE_KEY_SIZE_IN_BYTES) != 0) {
+	if (stm32_rng_read(seed, MCE_SEED_SIZE_IN_BYTES) != 0) {
 		panic();
 	}
 #else /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
 	if (stm32mp_is_wakeup_from_standby()) {
-		stm32mp1_pm_get_mce_mkey_from_context(mkey);
+		stm32mp1_pm_get_mce_seed_from_context(seed);
 		stm32_mce_reload_configuration();
 	} else {
-		/* Generate MCE master key from RNG */
-		if (stm32_rng_read(mkey, MCE_KEY_SIZE_IN_BYTES) != 0) {
+		/* Generate MCE master key seed from RNG */
+		if (stm32_rng_read(seed, MCE_SEED_SIZE_IN_BYTES) != 0) {
 			panic();
 		}
 
-		stm32mp1_pm_save_mce_mkey_in_context(mkey);
+		stm32mp1_pm_save_mce_seed_in_context(seed);
 	}
 #endif /* STM32MP_UART_PROGRAMMER || STM32MP_USB_PROGRAMMER */
 
+	if (generate_mce_key_from_seed(mkey, seed, MCE_SEED_SIZE_IN_BYTES) != 0) {
+		panic();
+	}
+
 	if (stm32_mce_write_master_key(mkey) != 0) {
 		panic();
 	}
diff --git a/plat/st/stm32mp1/include/stm32mp1_context.h b/plat/st/stm32mp1/include/stm32mp1_context.h
index 9b815023c9..42a00990a2 100644
--- a/plat/st/stm32mp1/include/stm32mp1_context.h
+++ b/plat/st/stm32mp1/include/stm32mp1_context.h
@@ -19,8 +19,8 @@ bool stm32_pm_context_is_valid(void);
 void stm32_restore_ddr_training_area(void);
 uint32_t stm32_pm_get_optee_ep(void);
 #if STM32MP13
-void stm32mp1_pm_save_mce_mkey_in_context(uint8_t *data);
-void stm32mp1_pm_get_mce_mkey_from_context(uint8_t *data);
+void stm32mp1_pm_save_mce_seed_in_context(uint8_t *data);
+void stm32mp1_pm_get_mce_seed_from_context(uint8_t *data);
 void stm32mp1_pm_save_mce_region(uint32_t index, struct stm32_mce_region_s *config);
 void stm32mp1_pm_get_mce_region(uint32_t index, struct stm32_mce_region_s *config);
 #endif
diff --git a/plat/st/stm32mp1/platform.mk b/plat/st/stm32mp1/platform.mk
index 8b9c135dc9..290839706f 100644
--- a/plat/st/stm32mp1/platform.mk
+++ b/plat/st/stm32mp1/platform.mk
@@ -235,7 +235,8 @@ PLAT_BL_COMMON_SOURCES  +=	drivers/st/nvmem/stm32mp_tamp_nvram_mp1.c
 ifeq ($(STM32MP13),1)
 PLAT_BL_COMMON_SOURCES	+=	drivers/st/clk/clk-stm32-core.c				\
 				drivers/st/clk/clk-stm32mp13.c				\
-				drivers/st/crypto/stm32_rng.c
+				drivers/st/crypto/stm32_rng.c				\
+				drivers/st/crypto/stm32_saes.c
 else
 PLAT_BL_COMMON_SOURCES	+=	drivers/st/clk/stm32mp1_clk.c
 endif
@@ -253,7 +254,6 @@ endif
 ifeq (${TRUSTED_BOARD_BOOT},1)
 ifeq ($(STM32MP13),1)
 BL2_SOURCES		+=	drivers/st/crypto/stm32_pka.c
-BL2_SOURCES		+=	drivers/st/crypto/stm32_saes.c
 endif
 endif
 
diff --git a/plat/st/stm32mp1/stm32mp1_context.c b/plat/st/stm32mp1/stm32mp1_context.c
index 877c435cf4..8684bdfe60 100644
--- a/plat/st/stm32mp1/stm32mp1_context.c
+++ b/plat/st/stm32mp1/stm32mp1_context.c
@@ -62,7 +62,7 @@ struct backup_data_s {
 	uint32_t bl2_code_end;
 	uint32_t bl2_end;
 #if STM32MP13
-	uint8_t mce_mkey[MCE_KEY_SIZE_IN_BYTES];
+	uint8_t mce_seed[MCE_SEED_SIZE_IN_BYTES];
 	struct stm32_mce_region_s mce_regions[MCE_IP_MAX_REGION_NB];
 #endif
 };
@@ -197,7 +197,7 @@ void stm32_restore_ddr_training_area(void)
 }
 
 #if STM32MP13
-void stm32mp1_pm_save_mce_mkey_in_context(uint8_t *data)
+void stm32mp1_pm_save_mce_seed_in_context(uint8_t *data)
 {
 	struct backup_data_s *backup_data;
 
@@ -209,12 +209,12 @@ void stm32mp1_pm_save_mce_mkey_in_context(uint8_t *data)
 
 	clk_enable(BKPSRAM);
 
-	memcpy(backup_data->mce_mkey, data, MCE_KEY_SIZE_IN_BYTES);
+	memcpy(backup_data->mce_seed, data, MCE_SEED_SIZE_IN_BYTES);
 
 	clk_disable(BKPSRAM);
 }
 
-void stm32mp1_pm_get_mce_mkey_from_context(uint8_t *data)
+void stm32mp1_pm_get_mce_seed_from_context(uint8_t *data)
 {
 	struct backup_data_s *backup_data;
 
@@ -227,7 +227,7 @@ void stm32mp1_pm_get_mce_mkey_from_context(uint8_t *data)
 
 	clk_enable(BKPSRAM);
 
-	memcpy(data, backup_data->mce_mkey, MCE_KEY_SIZE_IN_BYTES);
+	memcpy(data, backup_data->mce_seed, MCE_SEED_SIZE_IN_BYTES);
 
 	clk_disable(BKPSRAM);
 }
diff --git a/plat/st/stm32mp1/stm32mp1_def.h b/plat/st/stm32mp1/stm32mp1_def.h
index f19d07382b..7c38c4a738 100644
--- a/plat/st/stm32mp1/stm32mp1_def.h
+++ b/plat/st/stm32mp1/stm32mp1_def.h
@@ -611,6 +611,7 @@ enum ddr_type {
 #if STM32MP13
 #define MCE_BASE			U(0x58001000)
 #define MCE_KEY_SIZE_IN_BYTES		U(16)
+#define MCE_SEED_SIZE_IN_BYTES		U(4)
 #endif
 
 /*******************************************************************************
-- 
2.53.0

